libraries/lib-wave-track/CachingWideSampleSequence.cpp

```diff
@@ -0,0 +1,128 @@
+#include "CachingWideSampleSequence.h"
+#include "SampleTrackStuff.h"
+#include "WaveTrack.h"
+
+#include <cassert>
+
+namespace
+{
+void FillBufferFromTrackBlockSequence(
+   const LibSampleTrack::TrackBlockSequence& track, float* buffer,
+   long long bufferSize)
+{
+   auto i = 0u;
+   long long written = 0;
+   while (written < bufferSize)
+   {
+      if (i == track.clips.size())
+      {
+         const auto toWrite = bufferSize - written;
+         std::fill(buffer + written, buffer + written + toWrite, 0.f);
+         break;
+      }
+      auto& clip = track.clips[i];
+      auto toWrite =
+         std::min(clip.usefulLength.as_long_long(), bufferSize - written);
+      auto offset = clip.offsetIntoFirstBlock;
+      for (const auto& block : clip.blocks)
+      {
+         const auto toWriteFromBlock =
+            std::min<long long>(block->size() - offset, toWrite);
+         std::copy(
+            block->data() + offset, block->data() + offset + toWriteFromBlock,
+            buffer + written);
+         toWrite -= toWriteFromBlock;
+         written += toWriteFromBlock;
+         offset = 0;
+      }
+      ++i;
+      // TODO implement silence insertion
+   }
+}
+} // namespace
+
+CachingWideSampleSequence::CachingWideSampleSequence(const WaveTrack& waveTrack)
+    : mWaveTrack(waveTrack)
+{
+}
+
+size_t CachingWideSampleSequence::NChannels() const
+{
+   return mWaveTrack.NChannels();
+}
+
+const SampleTrack* CachingWideSampleSequence::GetSampleTrack() const
+{
+   return &mWaveTrack;
+}
+
+float CachingWideSampleSequence::GetChannelGain(int channel) const
+{
+   return mWaveTrack.GetChannelGain(channel);
+}
+
+size_t CachingWideSampleSequence::GetMaxBlockSize() const
+{
+   return mWaveTrack.GetMaxBlockSize();
+}
+
+sampleCount CachingWideSampleSequence::GetBlockStart(sampleCount t) const
+{
+   return mWaveTrack.GetBlockStart(t);
+}
+
+size_t CachingWideSampleSequence::GetBestBlockSize(sampleCount t) const
+{
+   return mWaveTrack.GetBestBlockSize(t);
+}
+
+bool CachingWideSampleSequence::Get(
+   size_t iChannel, size_t nBuffers, samplePtr buffers[], sampleFormat format,
+   sampleCount start, size_t len, fillFormat fill, bool mayThrow,
+   sampleCount* pNumWithinClips) const
+{
+   // For now leaving `GrabFloats` using t0 as query start time to minimize the
+   // diff.
+   auto t0 = mWaveTrack.LongSamplesToTime(start);
+   mCacheHolders = mWaveTrack.GrabFloats(iChannel, nBuffers, t0, len);
```

Comment:
This reassings `mCacheHolders`, recreating the complete cache stucture, each time `Get` is called.  Isn't this excessive?

Shouldn't you test whether the pointer is null first?

That work work assuming that sample sequence contents are unchanging during the lifetime of the decoartor.